pkg/generator/quay/quay.go

```diff
@@ -0,0 +1,203 @@
+	esmeta "github.com/external-secrets/external-secrets/apis/meta/v1"
+)
+
+type Generator struct {
+	httpClient *http.Client
+}
+
+const (
+	defaultQuayURL = "quay.io"
+
+	errNoSpec    = "no config spec provided"
+	errParseSpec = "unable to parse spec: %w"
+	errGetToken  = "unable to get authorization token: %w"
+
+	httpClientTimeout = 5 * time.Second
+)
+
+func (g *Generator) Generate(ctx context.Context, jsonSpec *apiextensions.JSON, kube client.Client, namespace string) (map[string][]byte, error) {
+	return g.generate(
+		ctx,
+		jsonSpec,
+		kube,
+		namespace,
+	)
+}
+
+func (g *Generator) generate(
+	ctx context.Context,
+	jsonSpec *apiextensions.JSON,
+	_ client.Client,
+	namespace string) (map[string][]byte, error) {
+	if jsonSpec == nil {
+		return nil, errors.New(errNoSpec)
+	}
+	res, err := parseSpec(jsonSpec.Raw)
+	if err != nil {
+		return nil, fmt.Errorf(errParseSpec, err)
+	}
+
+	// Fetch the service account token
+	token, err := fetchServiceAccountToken(ctx, res.Spec.ServiceAccountRef, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch service account token: %w", err)
+	}
+	url := res.Spec.URL
+	if url == "" {
+		url = defaultQuayURL
+	}
+
+	accessToken, err := getQuayRobotToken(token, res.Spec.RobotAccount, url, g.httpClient)
+	if err != nil {
+		return nil, err
+	}
+	exp, err := tokenExpiration(accessToken)
+	if err != nil {
+		return nil, err
+	}
+	return map[string][]byte{
+		"registry": []byte(url),
+		"auth":     []byte(b64.StdEncoding.EncodeToString([]byte(res.Spec.RobotAccount + ":" + accessToken))),
+		"expiry":   []byte(exp),
+	}, nil
+}
+
+func getClaims(tokenString string) (map[string]interface{}, error) {
+	// Split the token into its three parts
+	parts := strings.Split(tokenString, ".")
+	if len(parts) != 3 {
+		return nil, fmt.Errorf("invalid token format")
+	}
+
+	// Decode the payload (the second part of the token)
+	payload, err := b64.RawURLEncoding.DecodeString(parts[1])
+	if err != nil {
+		return nil, fmt.Errorf("error decoding payload: %w", err)
+	}
+
+	var claims map[string]interface{}
+	if err := json.Unmarshal(payload, &claims); err != nil {
+		return nil, fmt.Errorf("error un-marshaling claims: %w", err)
+	} else {
+		return claims, nil
+	}
+}
+
+func tokenExpiration(tokenString string) (string, error) {
+	claims, err := getClaims(tokenString)
+	if err != nil {
+		return "", fmt.Errorf("error getting claims: %w", err)
+	}
+	exp, ok := claims["exp"].(float64)
+	if ok {
+		return strconv.FormatFloat(exp, 'f', -1, 64), nil
+	} else {
+		return "", fmt.Errorf("exp claim not found or wrong type")
+	}
+}
+
+// https://docs.projectquay.io/manage_quay.html#exchanging-oauth2-robot-account-token
+func getQuayRobotToken(fedToken, robotAccount, url string, hc *http.Client) (string, error) {
+	if hc == nil {
+		hc = &http.Client{
+			Timeout: httpClientTimeout,
+		}
+	}
+
+	req, err := http.NewRequest("GET", "https://"+url+"/oauth2/federation/robot/token", http.NoBody)
```

Comment:
Make sure that `url` doesn't already have a scheme. It might be that people define one with scheme. Either strip it or set it, but make sure it's not `https://https://` :)